libr/io/section.c

```diff
@@ -1,619 +1,745 @@
-	return true;
+	return ret;
 }
 
-R_API void r_io_section_free(void *ptr) {
-	RIOSection *s = (RIOSection*)ptr;
-	if (s) {
-		free (s->name);
-		free (s);
+R_API bool r_io_section_bin_rm(RIO *io, ut32 bin_id) {
+	RIOSection *s;
+	SdbListIter *iter, *niter;
+	int length;
+	if (!io || !io->sections || !io->sections->head || !io->sec_ids) {
+		return false;
+	}
+	length = ls_length (io->sections);
+	ls_foreach_safe (io->sections, iter, niter, s) {
+		if (s->bin_id == bin_id) {
+			r_id_pool_kick_id (io->sec_ids, s->id);
+			ls_delete (io->sections, iter);
+		}
 	}
+	return (!(length == ls_length (io->sections)));
 }
 
-R_API void r_io_section_clear(RIO *io) {
-	r_list_free (io->sections);
-	io->sections = r_list_newf (r_io_section_free);
-	if (!io->sections) {
-		return;
+
+R_API RIOSection *r_io_section_get_name(RIO *io, const char *name) {
+	RIOSection *s;
+	SdbListIter *iter;
+	if (!io || !name || !io->sections) {
+		return NULL;
 	}
+	ls_foreach (io->sections, iter, s) {
+		if (s->name && (!strcmp (s->name, name))) {
+			return s;
+		}
+	}
+	return NULL;
 }
 
-// TODO: implement as callback
-R_API void r_io_section_list(RIO *io, ut64 offset, int rad) {
-	int i = 0;
-	RListIter *iter;
+R_API void r_io_section_cleanup(RIO *io) {
+	SdbListIter *iter, *ator;
 	RIOSection *s;
-
-	if (io->va || io->debug) {
-		offset = r_io_section_vaddr_to_maddr_try (io, offset);
-	}
-	// XXX - Should this print the section->id or the location in the
-	// rio sections array?
-	r_list_foreach (io->sections, iter, s) {
-		if (rad) {
-			char *n = strdup (s->name);
-			r_name_filter (n, strlen (n));
-			io->cb_printf ("f section.%s %"PFMT64d" 0x%"PFMT64x"\n",
-				n, s->size, s->vaddr);
-			io->cb_printf ("S 0x%08"PFMT64x" 0x%08"PFMT64x" 0x%08"
-				PFMT64x" 0x%08"PFMT64x" %s %s\n", s->paddr,
-				s->vaddr, s->size, s->vsize, n, r_str_rwx_i (s->flags));
-			if (s->arch && s->bits) {
-				const char *archName = r_sys_arch_str (s->arch);
-				io->cb_printf ("Sa %s %d 0x%08"PFMT64x"\n",
-					archName, s->bits, s->vaddr);
-			}
-			free (n);
+	if (!io || !io->sections || !io->sec_ids) {
+		return;
+	}
+	if (!io->files) {
+		r_io_section_fini (io);
+		r_io_section_init (io);
+		return;
+	}
+	ls_foreach_safe (io->sections, iter, ator, s) {
+		if (!s) {
+			ls_delete (io->sections, iter);
+		} else if (!r_io_desc_get (io, s->fd)) {
+			r_id_pool_kick_id (io->sec_ids, s->id);
+			ls_delete (io->sections, iter);
 		} else {
-			io->cb_printf ("[%02d] %c 0x%08"PFMT64x" %s va=0x%08"PFMT64x
-				" sz=0x%04"PFMT64x" vsz=0x%04"PFMT64x" %s",
-				s->id, (offset >= s->paddr && offset<s->paddr + s->size)?'*': '.',
-				s->paddr, r_str_rwx_i (s->flags), s->vaddr, s->size, s->vsize, s->name);
-			if (s->arch && s->bits) {
-				io->cb_printf ("  ; %s %d\n", r_sys_arch_str (s->arch), s->bits);
-			} else {
-				io->cb_printf ("\n");
+			if (s->filemap && !r_io_map_exists_for_id (io, s->filemap)) {
+				s->filemap = 0;
+			}
+			if (s->memmap && !r_io_map_exists_for_id (io, s->memmap)) {
```

Comment:
The function should check for uf already, so if first part befoew the && is not necessary